frame/lottery/src/lib.rs

```diff
@@ -0,0 +1,432 @@
+						} else {
+							// Else, kill the lottery storage.
+							*lottery = None;
+							return T::WeightInfo::on_initialize_end()
+						}
+						// We choose not need to kill Participants and Tickets to avoid a large number
+						// of writes at one time. Instead, data persists between lotteries, but is not used
+						// if it is not relevant.
+					}
+				}
+				return T::DbWeight::get().reads(1)
+			})
+		}
+	}
+}
+
+impl<T: Config> Module<T> {
+	/// The account ID of the lottery pot.
+	///
+	/// This actually does computation. If you need to keep using it, then make sure you cache the
+	/// value and only call this once.
+	pub fn account_id() -> T::AccountId {
+		T::ModuleId::get().into_account()
+	}
+
+	/// Return the pot account and amount of money in the pot.
+	// The existential deposit is not part of the pot so lottery account never gets deleted.
+	fn pot() -> (T::AccountId, BalanceOf<T>) {
+		let account_id = Self::account_id();
+		let balance = T::Currency::free_balance(&account_id)
+			.saturating_sub(T::Currency::minimum_balance());
+
+		(account_id, balance)
+	}
+
+	// Converts a vector of calls into a vector of call indices.
+	fn calls_to_indices(calls: &[<T as Config>::Call]) -> Result<Vec<CallIndex>, DispatchError> {
+		let mut indices = Vec::with_capacity(calls.len());
+		for c in calls.iter() {
+			let index = Self::call_to_index(c)?;
+			indices.push(index)
+		}
+		Ok(indices)
+	}
+
+	// Convert a call to it's call index by encoding the call and taking the first two bytes.
+	fn call_to_index(call: &<T as Config>::Call) -> Result<CallIndex, DispatchError> {
+		let encoded_call = call.encode();
+		if encoded_call.len() < 2 { Err(Error::<T>::EncodingFailed)? }
+		return Ok((encoded_call[0], encoded_call[1]))
+	}
+
+	// Logic for buying a ticket.
+	fn do_buy_ticket(caller: &T::AccountId, call: &<T as Config>::Call) -> DispatchResult {
+		// Check the call is valid lottery
+		let config = Lottery::<T>::get().ok_or(Error::<T>::NotConfigured)?;
+		let block_number = frame_system::Module::<T>::block_number();
+		ensure!(block_number < config.start.saturating_add(config.length), Error::<T>::AlreadyEnded);
+		ensure!(T::ValidateCall::validate_call(call), Error::<T>::InvalidCall);
+		let call_index = Self::call_to_index(call)?;
+		let ticket_count = TicketsCount::get();
+		let new_ticket_count = ticket_count.checked_add(1).ok_or(Error::<T>::Overflow)?;
+		// Try to update the participant status
+		Participants::<T>::try_mutate(&caller, |(lottery_index, participating_calls)| -> DispatchResult {
+			let index = LotteryIndex::get();
+			// If lottery index doesn't match, then reset participating calls and index.
+			if *lottery_index != index {
+				*participating_calls = Vec::new();
+				*lottery_index = index;
+			} else {
+				// Check that user is not already participating under this call.
+				ensure!(!participating_calls.iter().any(|c| call_index == *c), Error::<T>::AlreadyParticipating);
```

Comment:
I think the crate top documentation should say that only one ticket per call kind can be purchased.